cpp/mcap/include/mcap/mcap.hpp

```diff
@@ -108,4 +127,78 @@ struct UnknownRecord {
   mcap::ByteArray data;
 };
 
+struct McapWriterOptions {
+  bool indexed;
+  std::string profile;
+  std::string library;
+  mcap::KeyValueMap metadata;
+
+  McapWriterOptions(const std::string_view profile)
+      : indexed(false)
+      , profile(profile)
+      , library("libmcap " LIBRARY_VERSION) {}
+};
+
+class McapWriter {
+public:
+  ~McapWriter();
+
+  /**
+   * @brief Open a new MCAP file for writing and write the header.
+   *
+   * @param stream Output stream to write to.
+   */
+  void open(std::ostream& stream, const McapWriterOptions& options);
+
+  /**
+   * @brief Write the MCAP footer and close the output stream.
+   */
+  void close();
+
+  /**
+   * @brief Add channel info and set `info.channelId` to a generated channel id.
+   * The channel id is used when adding messages.
+   *
+   * @param info Description of the channel to register. The `channelId` value
+   *   is ignored and will be set to a generated channel id.
+   */
+  void registerChannel(mcap::ChannelInfo& info);
+
+  /**
+   * @brief Write a message to the output stream.
+   *
+   * @param msg Message to add.
+   * @return An error code on failure.
+   */
+  std::optional<std::error_code> write(const mcap::Message& message);
+
+  /**
+   * @brief Write an attachment to the output stream.
+   *
+   * @param attachment Attachment to add.
+   * @return An error code on failure.
+   */
+  std::optional<std::error_code> write(const mcap::Attachment& attachment);
+
+private:
+  std::ostream* stream_ = nullptr;
```

Comment:
I hate trailing `_` but I'll shut up, it's recommended in the google style guide 😭